File: soft/computer/plume_api/python/test_plume_api.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import sys
import numpy as np
import time
import signal

from plume_api import *

def _callback(type, MAC):
    global plumeApi

    plumeReceiver = plumeApi.getPlumeReceiverFromMAC(MAC)

    if type == TYPE_DOF:
        print("new DOF " + str(plumeReceiver.getMACHex()))
        position = np.zeros(3, dtype=np.double)
        orientation = np.zeros(3, dtype=np.double)
        plumeReceiver.get6DOFDeg(position, orientation)
        print(position)
        print(orientation)
    elif type == TYPE_BATTERY:
        print("new battery " + str(plumeReceiver.getMACHex()) + " " + str(plumeReceiver.getBattery()))
    elif type == TYPE_STATE:
        print("new state " + str(plumeReceiver.getMACHex()) + " " + str(plumeReceiver.getState()))

callback = callback_t(_callback)

def signalHandler(signal, frame):
    global plumeApi

    print("You pressed Ctrl+C - or killed me with -2")

    del plumeApi

    sys.exit(0)

def init():
    global plumeApi

    print("test plume API")

    plumeApi = PlumeApi("../../calculator/calibration.json")

    signal.signal(signal.SIGINT, signalHandler)

    plumeEmitter0 = plumeApi.addEmitter("1", "/dev/ttyACM0")

    plumeApi.setFrequency(50)
    plumeApi.setCallback(callback)
    plumeApi.scanDevice()

    # Wait one receiver
    while plumeApi.getPlumeReceiversSize() < 1:
        time.sleep(0.01)

    plumeApi.startEmitters()
    plumeApi.getPlumeReceiversAt(0).start();

    while True:
        for i in range(plumeApi.getPlumeReceiversSize()):
            plumeReceiver = plumeApi.getPlumeReceiversAt(i)
            print(plumeReceiver.getMACHex())
            position = np.zeros(3, dtype=np.double)
            orientation = np.zeros(3, dtype=np.double)
            plumeReceiver.get6DOFDeg(position, orientation)
            print(position)

        time.sleep(0.5)

init()

File: soft/computer/plume_api/C/test_plume_api.cpp
#include <iostream>
#include <unistd.h>
#include <signal.h>

#include "plume_api.h"

using namespace std;

void callback(int type, PlumeReceiver* plumeReceiver){
    printf("-----\n\r");

    if(type == TYPE_DOF){
        double position[3];
        double orientation[3];

        plumeReceiver->get6DOFDeg(position, orientation);

        printf("new DOF ");
        cout << plumeReceiver->MAC << endl;
        printf("position %f, %f, %f\n\r", position[0], position[1], position[2]);
        printf("orientation %f, %f, %f\n\r", orientation[0], orientation[1], orientation[2]);
    }
    else if(type == TYPE_BATTERY){
        printf("new battery ");
        cout << plumeReceiver->MAC;
        printf(" %u\n\r", plumeReceiver->battery);
    }
    else if(type == TYPE_STATE){
        printf("new state ");
        cout << plumeReceiver->MAC;
        printf(" %u\n\r", plumeReceiver->state);
    }
}

static PlumeApi* plumeApi;

void signalHandler(int sig){
    (void) sig;

    printf("You pressed Ctrl+C - or killed me with -2\n");

    delete plumeApi;

    exit(0);
}

int main(int argc, char **argv){
    (void) argc;
    (void) argv;

    using namespace std::placeholders;

    cout << "test plume API" << endl;

    plumeApi = new PlumeApi("../../calculator/calibration.json");

    signal(SIGINT, signalHandler);

    PlumeEmitter* plumeEmitter0 = plumeApi->addEmitter("1", "/dev/ttyACM0");

    plumeApi->setFrequency(50);
    plumeApi->setCallback(bind(callback, _1, _2));
    plumeApi->scanDevice();

    // Wait one receiver
    while(plumeApi->plumeReceivers.size() < 1){
        usleep(5000);
    }

    plumeApi->startEmitters();
    plumeApi->plumeReceivers[0]->start();

    while(1){
        unsigned i,j;
        for(i = 0, j = plumeApi->plumeReceivers.size(); i < j; i++){
            cout << plumeApi->plumeReceivers[i]->MAC << endl;
        }
        usleep(50000);
    }

    return 0;
}